Userland/Libraries/LibWeb/Layout/TreeBuilder.cpp

```diff
@@ -733,4 +734,50 @@ void TreeBuilder::missing_cells_fixup(Vector<JS::Handle<Box>> const& table_root_
         });
     }
 }
+
+static bool is_row_group(Node const& node)
+{
+    // NB: For the purpose of reordering row groups, rows can be treated as a group with one element.
+    return node.display().is_table_header_group() || node.display().is_table_row_group() || node.display().is_table_footer_group() || node.display().is_table_row();
+}
+
+void TreeBuilder::reorder_row_groups(NodeWithStyle& root)
+{
+    for_each_in_tree_with_inside_display<CSS::Display::Inside::Table>(root, [&](auto& parent) {
+        Vector<JS::Handle<Node>> row_groups;
+        // Per https://www.w3.org/TR/css-tables-3/#table-header-group:
+        // Like table-row-group but, for layout purposes, the first such row group is always displayed before all other rows and row groups.
+        const Node* header_group { nullptr };
+        for (auto child = parent.first_child(); child; child = child->next_sibling())
+            if (child->display().is_table_header_group()) {
+                row_groups.append(child);
+                header_group = child;
+                break;
+            }
+        JS::Handle<Node> footer_group;
+        for (auto child = parent.first_child(); child; child = child->next_sibling()) {
+            if (!is_row_group(*child))
+                continue;
+            if (child->display().is_table_footer_group() && !footer_group)
+                footer_group = child;
+            // Accumulate table row groups and header / footer groups which aren't the first and behave like regular row groups.
+            else if (child != header_group)
+                row_groups.append(child);
+        }
+        // Per https://www.w3.org/TR/css-tables-3/#table-footer-group:
+        // Like table-row-group but, for layout purposes, the first such row group is always displayed after all other rows and row groups.
+        if (footer_group)
+            row_groups.append(footer_group);
+        Vector<JS::Handle<Node>> sequence;
+        for (auto child = parent.first_child(); child; child = child->next_sibling())
+            if (is_row_group(*child))
+                sequence.append(*row_groups.take_first());
+            else
+                sequence.append(*child);
```

Comment:
as I understand `sequence` is supposed to contain all parent's children when loop is done. why `row_groups.take_first()` happens for row groups? can we always do `sequence.append(*child)` instead?

Reply:
The specification doesn't say we should reorder any other children except for rows, row groups and alike (header groups and footer groups). Going by it, we have to preserve the relative order of everything else.